frontends/p4/moveDeclarations.cpp

```diff
@@ -109,24 +109,41 @@ const IR::Node *MoveInitializers::preorder(IR::P4Parser *parser) {
             }
         }
     }
+
     if (someInitializers) {
+        // Check whether any of the parser's states loop back to the start state.
+        // If this is the case, then the parser's decl initializers cannot be moved
+        // to the start state. A new start state that is never looped back to must
+        // be inserted at the head of the graph.
+        for (const auto *state : parser->states) {
```

Comment:
This probably could be a self-contained helper function which returns true if a loop is detected